File: pascal/stack.pas
unit stack;

interface

type
    TNode = record
        element: integer;
        next: ^TNode;
    end;

    TStack = record
        size: integer;
        top: ^TNode;
    end;

    PStack = ^TStack;

function new(): TStack;
procedure push(stack: PStack; element: integer);
function pop(stack: PStack): integer;
function peek(stack: PStack): integer;

implementation

function new(): TStack;
begin
    new.size := 0;
    new.top := nil;
end;

procedure push(stack: PStack; element: integer);

var
    newNode: ^TNode;

begin
    newNode := GetMem(sizeOf(TNode));

    newNode^.element := element;
    newNode^.next := stack^.top;

    stack^.top := newNode;
    stack^.size += 1;
end;

function pop(stack: PStack): integer;

var
    auxElem: integer;
    auxPtr: ^TNode;

begin
    if stack^.top = nil then
        exit(0)
    else
    begin
        auxElem := stack^.top^.element;
        auxPtr := stack^.top^.next;

        freeMem(stack^.top);

        stack^.top := auxPtr;
        stack^.size -= 1;

        pop := auxElem;
    end;
end;    

function peek(stack: PStack): integer;
begin
    if stack^.top = nil then
        exit(0)
    else
        exit(stack^.top^.element);
end;

procedure print(s: PStack);

var
    auxPtr: ^TNode;

begin
    auxPtr := s^.top;

    while auxPtr <> nil do
    begin
        writeLn(auxPtr^.element);
        auxPtr := auxPtr^.next;
    end;
    writeLn('-----');
end;

end.

File: rust/src/stack.rs
struct Node {
    element: i32,
    next: Option<Box<Node>>
}

pub struct Stack {
    pub size: i32,
    top: Option<Box<Node>>
}

impl Stack {
    pub fn new() -> Self {
        Stack {
            size: 0,
            top: None
        }
    }

    pub fn push(&mut self, element: i32) {
        let new_node = Box::new(Node{
            element: element,
            next: self.top.take()
        });

        self.top = Some(new_node);
        self.size += 1;
    }

    pub fn pop(&mut self) -> i32 {
        match self.top.take() {
            None => {
                return 0;
            }
            
            Some(node) => {
                self.top = node.next;
                self.size -= 1;
                
                return node.element;
            }
        }
    }

    pub fn peek(&self) -> i32 {
        match &self.top {
            None => {
                return 0;
            }

            Some(n) => {
                return n.element;
            }
        }
    }
}

#[cfg(test)]
mod test {
    #[test]
    fn basics() {
        use crate::stack::Stack;

        let mut list = Stack::new();

        // Check empty list behaves right
        assert_eq!(list.pop(), 0);

        // Populate list
        list.push(1);
        list.push(2);
        list.push(3);

        // Check normal removal
        assert_eq!(list.pop(), 3);
        assert_eq!(list.pop(), 2);

        // Push some more just to make sure nothing's corrupted
        list.push(4);
        list.push(5);

        // Check normal removal
        assert_eq!(list.pop(), 5);
        assert_eq!(list.pop(), 4);

        // Check exhaustion
        assert_eq!(list.pop(), 1);
        assert_eq!(list.pop(), 0);
    }
}
